File: jobs/go/http/job-service/job-service.go
/*
Copyright 2021 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
     http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
)

type Job struct {
	TypeURL string `json:"type_url"`
	Value   string `json:"value"`
}

type DroidJob struct {
	Droid string `json:"droid"`
	Task  string `json:"task"`
}

func main() {
	appPort := os.Getenv("APP_PORT")
	if appPort == "" {
		appPort = "6200"
	}

	// Setup job handler
	http.HandleFunc("/job/", handleJob)

	fmt.Printf("Server started on port %v\n", appPort)
	err := http.ListenAndServe(":"+appPort, nil)
	if err != nil {
		log.Fatal(err)
	}

}

func handleJob(w http.ResponseWriter, r *http.Request) {
	fmt.Println("Received job request...")
	rawBody, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, fmt.Sprintf("error reading request body: %v", err), http.StatusBadRequest)
		return
	}

	var jobData Job
	if err := json.Unmarshal(rawBody, &jobData); err != nil {
		http.Error(w, fmt.Sprintf("error decoding JSON: %v", err), http.StatusBadRequest)
		return
	}

	// Decoding job data
	decodedValue, err := base64.RawStdEncoding.DecodeString(jobData.Value)
	if err != nil {
		fmt.Printf("Error decoding base64: %v", err)
		http.Error(w, fmt.Sprintf("error decoding base64: %v", err), http.StatusBadRequest)
		return
	}

	// Creating Droid Job from decoded value
	droidJob := setDroidJob(string(decodedValue))

	fmt.Println("Starting droid:", droidJob.Droid)
	fmt.Println("Executing maintenance job:", droidJob.Task)

	w.WriteHeader(http.StatusOK)
}

func setDroidJob(decodedValue string) DroidJob {
	// Removing new lines from decoded value - Workaround for base64 encoding issue
	droidStr := strings.ReplaceAll(decodedValue, "\n", "")
	droidArray := strings.Split(droidStr, ":")

	droidJob := DroidJob{Droid: droidArray[0], Task: droidArray[1]}
	return droidJob
}

File: jobs/python/http/job-service/job-service.py
import os
from flask import Flask, request, jsonify
import json
import traceback

app = Flask(__name__)

class DroidJob:
    def __init__(self, droid: str, task: str):
        self.droid = droid
        self.task = task

def set_droid_job(decoded_value: str) -> DroidJob:
    # Remove newlines from decoded value and split into droid and task
    droid_str = decoded_value.replace('\n', '')
    droid_array = droid_str.split(':')
    return DroidJob(droid_array[0], droid_array[1])

@app.route('/job/<job>', methods=['POST'])
def handle_job(job):
    print(f"Received job request...")
    
    try:
        print("Raw request data:", request.get_data().decode('utf-8'))
        job_data = request.get_json()
        print("Parsed job data:", json.dumps(job_data, indent=2))
        
        if not job_data:
            return "Error reading request body", 400
        
        # In Dapr, the job data comes in a special "data" field within the request
        value = job_data.get('data', {}).get('value', '')
        print(f"Extracted value: {value}")
            
        # Create DroidJob from value
        droid_job = set_droid_job(value)
        
        print(f"Starting droid: {droid_job.droid}")
        print(f"Executing maintenance job: {droid_job.task}")
        
        return "", 200
        
    except Exception as e:
        print(f"Error processing job request:")
        print(traceback.format_exc())
        return f"Error processing job: {str(e)}", 400

if __name__ == '__main__':
    app_port = os.getenv('APP_PORT', '6200')
    print(f"Server started on port {app_port}")
    app.run(host='0.0.0.0', port=int(app_port))